### ISSUE TITLE
"list[int]" is incompatible with "Iterable[SupportsAbs[_T@abs]]"

### ISSUE THREAD

**Issue by niklasekstrom** [microsoft/pyright#4281]
**Describe the bug**
Using the following short program:
```python
xs = [1, -2, 3]
ys = map(abs, xs)
print(list(ys))
```

The Pylance type checker (I'm running VSCode) complains and says:
```
Argument of type "list[int]" cannot be assigned to parameter "__iter1" of type "Iterable[_T1@__init__]" in function "__init__"
  "list[int]" is incompatible with "Iterable[SupportsAbs[_T@abs]]"
    TypeVar "_T_co@Iterable" is covariant
      "int" is incompatible with protocol "SupportsAbs[_T@abs]"
        TypeVar "_T_co@SupportsAbs" is covariant
          Type "int" cannot be assigned to type "_T@abs"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

**To Reproduce**
As above.

**Expected behavior**
I expect the type checker to happily accept the above program without complaints, or otherwise explain in a clear way why the program is badly typed.

**Comment by erictraut** [comment]
I think this is correct behavior. Mypy produces the same error.

I'm going to leave the bug open though because I want to investigate further and give you a full explanation.

The example is obscured by complex overloads. Here's a simplified version:

```python
from typing import Any, Callable, Iterable, SupportsAbs, TypeVar

_T = TypeVar("_T")

xs: Iterable[int] = [1, -2, 3]

def my_map(a: Callable[[_T], Any], b: Iterable[_T]): ...
def my_abs(__x: SupportsAbs[_T]) -> _T: ...

ys = my_map(my_abs, xs)
```

**Comment by Azureblade3808** [comment]
Passing arguments in different orders yields different results, which I don't think is quite right --
```python
from typing import Callable, Protocol, TypeVar

_T = TypeVar("_T")
_T_co = TypeVar("_T_co", covariant=True)
_U = TypeVar("_U")

class MyIterable(Protocol[_T_co]): ...
class MySupportsAbs(Protocol[_T_co]): ...
def my_abs(x: MySupportsAbs[_T], /) -> _T: ...

# Using named arguments instead of pos-only ones.
def my_map(a: Callable[[_T], _U], b: MyIterable[_T]) -> MyIterable[_U]: ...

xs: MyIterable[MySupportsAbs[int]] = globals()[""]
reveal_type(xs)

ys0 = my_map(a=my_abs, b=xs)  # Failed with normal order of arguments.
reveal_type(ys0)

ys1 = my_map(b=xs, a=my_abs)  # Passed with reversed order of arguments.
reveal_type(ys1)
```

**Comment by erictraut** [comment]
I agree that order dependence isn't ideal, but the constraint solver necessarily involves many heuristics, and these can result in different behaviors depending on order. I'll investigate whether there's a way to make this non-order-dependent without breaking other cases.

**Comment by niklasekstrom** [comment]
> I think this is correct behavior. Mypy produces the same error.

When you say that the behavior is correct, what does that mean? Does it mean that the program I wrote above is incorrect, or does it mean that the correct behavior for pyright is to flag a correct program as having typing issues?

Is there some way that I could have structured my program to avoid the typing warnings? Or what do you suggest that I should do in my program to avoid this warning?

Thanks!

**Comment by erictraut** [comment]
I've enhanced pyright's constraint solver so it can handle the case where a generic callback function can be passed as an argument to a generic higher-order function. Pyright can now solve the type variables in both the callback and the callee in a deterministic manner. This will be included in the next release.

Note that mypy doesn't support this functionality, so if you plan to use both mypy and pyright to type check your code base, this might be an issue.

**Comment by erictraut** [comment]
This is included in pyright 1.1.286, which I just published. It will also be included in a future version of pylance.
